pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/DimensionTableDataManager.java

```diff
@@ -0,0 +1,162 @@
+  }
+
+  public static DimensionTableDataManager getInstanceByTableName(String tableNameWithType) {
+    return _instances.get(tableNameWithType);
+  }
+
+  /**
+   * Instance properties/methods
+   */
+  private final ReadWriteLock _rwl = new ReentrantReadWriteLock();
+  private final Lock _lookupTableReadLock = _rwl.readLock();
+  private final Lock _lookupTableWriteLock = _rwl.writeLock();
+
+  // _lookupTable is a HashMap used for storing/serving records for a table keyed by table PK
+  @GuardedBy("_rwl")
+  private final Map<PrimaryKey, GenericRow> _lookupTable = new HashMap<>();
+  private Schema _tableSchema;
+  private List<String> _primaryKeyColumns;
+
+  @Override
+  protected void doInit() {
+    super.doInit();
+
+    _tableSchema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkState(_tableSchema != null, "Failed to find schema for table: %s", _tableNameWithType);
+    _primaryKeyColumns = _tableSchema.getPrimaryKeyColumns();
+    Preconditions.checkState(!CollectionUtils.isEmpty(_primaryKeyColumns),
+        "Primary key columns must be configured for dimension table: %s", _tableNameWithType);
+  }
+
+  @Override
+  public void addSegment(File indexDir, IndexLoadingConfig indexLoadingConfig)
+      throws Exception {
+    super.addSegment(indexDir, indexLoadingConfig);
+    try {
+      loadLookupTable();
+      _logger.info("Successfully loaded lookup table: {}", getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(String.format("Error loading lookup table: %s", getTableName()), e);
+    }
+  }
+
+  @Override
+  public void removeSegment(String segmentName) {
+    super.removeSegment(segmentName);
+    try {
+      loadLookupTable();
+      _logger.info("Successfully removed segment and reloaded lookup table: {}", getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Error reloading lookup table after segment remove for table: {}", getTableName()), e);
+    }
+  }
+
+  /**
+   * `loadLookupTable()` reads contents of the DimensionTable into _lookupTable HashMap for fast lookup.
+   */
+  private void loadLookupTable()
+      throws Exception {
+    _lookupTableWriteLock.lock();
+    try {
+      _lookupTable.clear();
+      List<SegmentDataManager> segmentManagers = acquireAllSegments();
+      if (segmentManagers.size() == 0) {
+        return;
+      }
+
+      for (SegmentDataManager segmentManager : segmentManagers) {
+        IndexSegment indexSegment = segmentManager.getSegment();
+        try (PinotSegmentRecordReader reader =
+            new PinotSegmentRecordReader(indexSegment.getSegmentMetadata().getIndexDir())) {
+          while (reader.hasNext()) {
+            GenericRow row = reader.next();
+            _lookupTable.put(row.getPrimaryKey(_primaryKeyColumns), row);
+          }
+        }
+        releaseSegment(segmentManager);
```

Comment:
We might want to put release in the finally block so that it is always released even with exceptions:
```suggestion
  try {
    for (SegmentDataManager segmentManager : segmentManagers) {
      ...
    }
  } finally {
    for (SegmentDataManager segmentManager : segmentManagers) {
      releaseSegment(segmentManager);
    }
  }
```

Reply:
Good suggestion. How about creating the try..finally block inside the for loop like:
```java
for (SegmentDataManager segmentManager: segmentManagers) {
  try {
     //... read contents...
  } finally {
    releaseSegment(segmentManager);
  }
}
```

Reply:
This won't work because if the loading throws exception, it will only release one segment manager

Reply:
Oh, you're right, missed that completely. Updating as suggested.